### DISCUSSION TITLE
Using GraphQL API in getStaticProps or getServerSideProps

### DISCUSSION THREAD

**Discussion by matepapp** [vercel/next.js#10946] [Help]
Hey! 👋 

Thanks for releasing NextJS 9.3! Although it's _only_ a minor version bump, it's definitely a **major** step forward! 

Unfortunately, I couldn't find any example/docs which demonstrates how to call a GraphQL API either in `getStaticProps` or `getServerSideProps`. This could be using one of the popular clients such as [Apollo Client](https://github.com/apollographql/apollo-client), [urql](https://github.com/FormidableLabs/urql), [React Query](https://github.com/tannerlinsley/react-query), etc.. Currently we're using the [next-with-apollo](https://github.com/lfades/next-with-apollo) package which (as far as I know) basically wraps `getInitialProps` and enables to use `useQuery` or `useMutation` hooks in your components. Is there an equivalent solution or workaround which supports the newly released APIs?

**Comment by timneutkens** [comment]
Here's a minimal example for fetching a GraphQL API: https://github.com/zeit/next.js/blob/canary/examples/cms-datocms/lib/api.js#L23-L42

The way that Apollo does tree-rendering to allow useQuery causes quite a lot of performance overhead (and general bundle size when using Apollo is quite heavy currently). So would recommend using a lightweight solution.

Using the plain Apollo client without `getDataFromTree` is fine though.

**Reply by dhazeley** [reply]
@timneutkens Hey, I'm curious about why Apollo's useQuery has a lot of performance overhead. In what cases do you think the performance overhead would be worth it to use Apollo client?

Also if you could point me to any resources I can read up on that would be helpful. I'm evaluating whether I should go with a next-with-apollo approach, or the more lightweight approach you just mentioned.

**Reply by timneutkens** [reply]
Code can be found here: https://github.com/apollographql/react-apollo/blob/master/packages/ssr/src/getDataFromTree.ts#L23-L54

https://github.com/apollographql/react-apollo/blob/master/packages/ssr/src/getDataFromTree.ts#L13

First of all note that the following explanation is unrelated to Next.js.

The way that `getDataFromTree` sort of works is:

`getDataFromTree` is called with a React tree, eg `getDataFromTree(AppTree)`

It'll run `renderToStaticMarkup ` from ReactDOMServer. Note that this is the React SSR API and means that it does a full server-render of the whole React tree.

Note that `renderToStaticMarkup` is a synchronous run to completion method, meaning that it can't await promises as of right now (Suspense might solve this).

In practice though you have `useQuery` and `<Query>` components deeply nested in the React tree. React can't await those as said, so this is worked around by throwing a promise every time a query is found.

When the promise is thrown that is `await`ed and then the rendering starts again, from the beginning of the tree.

This means that if you have nested queries you cause a lot of full server-renders.

**Reply by janus-reith** [reply]
@timneutkens So the usecase you are referring to is, if you use `getDataFromTree` on the server if you need to make sure all your `useQuery` hooks or <Query> components within your codebase would actually hold data for the static page, right?
Besides maybe bundle size, there would be no concern of using `useQuery` in general on the client, while finding other options for fetching the data for the initial static page, or simply omitting that data if not releveant(e.g. authentication related).

I realize you already explained above how this is related to `getDataFromTree`, but wanted to underline that, as someone who quickly glimpses over this discussion could get the wrong impression about `useQuery`.

**Comment by Manubi** [comment]
Hi,
another approach with grapqhl-request. Maybe it's interesting for the people who use swr with graphql-request. 
```
import { Layout } from "@components/layout";
import { API } from "@graphQLClient";
import { VACCINATION_ID, VACCINATION_PAGE } from "@impfi.query";
import { request } from "graphql-request";
import { GetStaticProps, NextPage } from "next";
import React from "react";

interface IVaccinationId {
    id: string;
}

interface IVaccinationQuery {
    vaccination: IVaccinationId[];
}
interface IVaccination {
    id: string;
    name: string;
    description: string;
}

interface IProps {
    data: {
        vaccination: IVaccination[];
    };
}

const Vaccines: NextPage<IProps> = (props) => {
    const vaccination = props.data.vaccination[0];
    return (
        <Layout>
            <div className="hero m-2">
                <h1 className="text-gray-800 text-xl">{vaccination.name}</h1>
                <p>{vaccination.description}</p>
            </div>
        </Layout>
    );
};

export const getStaticPaths = async () => {
    const res: IVaccinationQuery = await request(API, VACCINATION_ID);
    const paths = res.vaccination.map((vacc) => ({ params: { id: vacc.id } }));
    return { paths, fallback: false };
};

export const getStaticProps: GetStaticProps<IProps> = async ({ params }) => {
    const data = await request(API, VACCINATION_PAGE, {
        id: params?.id
    });
    return { props: { data } };
};

export default Vaccines;

```

**Comment by githubjosh** [comment]
Anyone figure this out with Apollo? Specifically, SSR with client side querying of the Apollo store: e.g. loading posts (server side with getServerSideProps) with client side paging.

@timneutkens I checked out the example above, but to your point, it doesn't really fit the use case. The example makes fetch requests, then relies on props drilling versus reading from and updating the Apollo store.

This one's been killin' me.

**Reply by SoorajChandran** [reply]
Stumbled upon this with the same issue. Since I'm a next js beginner, I thought I was missing something. This comment is an affirmation for me. 

Having said that, can I use `getInitialProps` to server render the content for SEO?
Also, is there a way to use hooks inside `getInitialProps` and fetch data from Graphql?

My fallback is that I'll have to use REST to fetch data in `getServerSideProps` :/ Which is not nice at all.

**Reply by johnpaul89** [reply]
> Anyone figure this out with Apollo? Specifically, SSR with client side querying of the Apollo store: e.g. loading posts (server side with getServerSideProps) with client side paging.
> 
> @timneutkens I checked out the example above, but to your point, it doesn't really fit the use case. The example makes fetch requests, then relies on props drilling versus reading from and updating the Apollo store.
> 
> This one's been killin' me.

Check their official example it solves all the problems https://github.com/zeit/next.js/tree/canary/examples/with-apollo

**Reply by SoorajChandran** [reply]
Thanks! I made it work with this approach. Left a comment on a similar discussion

https://github.com/lfades/next-with-apollo/issues/114#issuecomment-622082403

**Comment by AryanJ-NYC** [comment]
```
import { gql, ApolloClient, HttpLink, InMemoryCache } from '@apollo/client';
import fetch from 'isomorphic-unfetch';
import { GetServerSideProps } from 'next';

export const getApolloClient = () => {
  const isLocalHost = process.env.NOW_URL?.includes('localhost');
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      uri: `http${isLocalHost ? '' : 's'}://${process.env.NOW_URL}/api/graphql`,
      fetch,
    }),
  });
};

const TRANSACTIONS_BY_PRODUCT_ID_QUERY = gql`
  query TRANSACTIONS_BY_PRODUCT_ID_QUERY($product_id: String!) {
    transactionsByProductId(productId: $product_id) {
      id
      bitcoinAddress
      blockchainTxId
      invoice {
        status
      }
    }
  }
`;
export const getServerSideProps: GetServerSideProps = async ({ params }) => {
  const apolloClient = getApolloClient();
  const { id } = params;
  const { data } = await apolloClient.query({
    query: TRANSACTIONS_BY_PRODUCT_ID_QUERY,
    variables: { product_id: id },
  });
  const transactions: Transaction[] = data.transactionsByProductId;
  return { props: { transactions } };
};
```

works for me.

**Reply by spwr007** [reply]
This is great, I have been implementing this way, but it is working for a authorised ApolloClient with headers in serverSiderProps.
Is there any way to solve it?

**Reply by AryanJ-NYC** [reply]
I wish I knew. Since posting my solution above, I've [dropped Apollo Client altogether and moved to a simpler approach](https://dev.to/aryanjnyc/i-migrated-away-from-apollo-client-to-vercel-swr-and-prisma-graphql-request-and-you-can-too-245b):

```typescript
import { GetServerSideProps } from "next";
import { GraphQLClient } from "graphql-request";

const graphQlClient =  new GraphQLClient(`${process.env.BASE_URL}/api/graphql`)

const TRANSACTIONS_BY_PRODUCT_ID_QUERY = /* GraphQL */ `
  query TRANSACTIONS_BY_PRODUCT_ID_QUERY($product_id: String!) {
    transactionsByProductId(productId: $product_id) {
      id
      bitcoinAddress
      blockchainTxId
      invoice {
        status
      }
    }
  }
`;
export const getServerSideProps: GetServerSideProps = async ({ params }) => {
  // @ts-expect-error
  const { id } = params;

  const data: {
    transactionsByProductId: Transaction[];
  } = await graphQlClient.request(TRANSACTIONS_BY_PRODUCT_ID_QUERY, {
    product_id: id,
  });

  const transactions = data.transactionsByProductId;
  return {
    props: {
      transactions,
    },
  };
};
```

Unsure how you'd add authorization headers here without making the call from the client, though.

**Reply by Manubi** [reply]
Hey, 
this should work. Just add the header after the endpoint.

```
  const endpoint = 'https://api.graph.cool/simple/v1/cixos23120m0n0173veiiwrjr'

  const graphQLClient = new GraphQLClient(endpoint, {
    headers: {
      authorization: 'Bearer MY_TOKEN',
    },
  })

```

**Reply by AryanJ-NYC** [reply]
Right. My question is more: how would we get that token that's likely stored in the browser?

Since that's likely impossible as this code is run on the server-side, I think client-side fetching (if we need that `MY_TOKEN`) is the way to go.

**Reply by Manubi** [reply]
That depends on how you store your token. If you use sessions you can have a look here:
https://github.com/vercel/next.js/blob/canary/examples/auth0/pages/advanced/ssr-profile.js

**Reply by balazsorban44** [reply]
What I do in my Oauth2 conformant app is that I create a service client, and using a [Client Credential Flow](https://auth0.com/docs/flows/concepts/client-credentials) to obtain an access token, before creating my apollo client. 

Since you are on the server-side, you can be a bit less strict, and attain a powerful token for your API, either by adding it as an environment variable through .env, or make a call to your authorization server and ask for one.

**Reply by spwr007** [reply]
@AryanJ-NYC Thank you for the examples. They are really great. It made the code look great and easy to implement.
Also about the query, how we can do that on server side. I created a NextJs API end point which sets the header cookie and we can use the cookie in serverSideProps. This solved my problem.

**Reply by ITSWYoo** [reply]
@AryanJ-NYC 
  hi
i use graphql-code-generator
do you know how to generate this code to typescript code ?
```js
  const { data } = await apolloClient.query({
    query: TRANSACTIONS_BY_PRODUCT_ID_QUERY,
    variables: { product_id: id },
  });
```
i want to query, variables type checking

**Reply by warent** [reply]
Thanks @AryanJ-NYC & @Manubi :) 👍

**Reply by smblee** [reply]
@ITSWYoo 

`apolloClient.query` takes in generic types for res/var.

```
  const { data } = await apolloClient.query<TransactionsByProductIdQuery, TransactionsByProductIdQueryVariables>({
    query: TRANSACTIONS_BY_PRODUCT_ID_QUERY,
    variables: { product_id: id },
  });
```

I get these autogenerated using the `codegen.js` file.

```
module.exports = {
  schema: [
    {
      'http://localhost:3030/v1/graphql': {
        headers: {
          'Authorization': 'bearer 12345',
        },
      },
    },
  ],
  documents: ['./src/**/*.tsx', './src/**/*.ts', './src/**/*.graphql'],
  overwrite: true,
  generates: {
    './src/graphql/generated/index.tsx': {
      plugins: ['typescript', 'typescript-operations', 'typescript-react-apollo'],
      config: {
        skipTypename: false,
        withHooks: true,
        withHOC: false,
        withComponent: false,
      },
    },
    './schema/graphql.schema.json': {
      plugins: ['introspection'],
    },
  },
  hooks: {
    afterAllFileWrite: ['prettier --write'],
  },
};
```

**Reply by balazsorban44** [reply]
https://graphql-code-generator.com/docs/plugins/typescript-react-apollo

**Comment by benjaminpearson** [comment]
Here is an example https://github.com/vercel/next.js/pull/13202 if you want to query your API route GraphQL endpoint in order to do Static Generation with `getStaticProps` and `getStaticPaths`. The example may be helpful to do the server / static rendering aspect, you would still need a graphql client if you needed to query client side, but the examples folder has quite a few that illustrate options available https://github.com/vercel/next.js/tree/canary/examples

**Comment by jdponomarev** [comment]
I had a similar issue, was getting "XMLHttpRequest is not defined". My problem was that I was using unfetch instead of fetch and didn't remove it in getStaticProps
